routing/probability_bimodal.go

```diff
@@ -0,0 +1,432 @@
+// to the upper value.
+func (p *probabilityBimodal) integral(capacity, lower, upper float64) float64 {
+	if lower < 0 || lower > upper {
+		log.Errorf("probability integral limits nonsensical: capacity:"+
+			"%v lower: %v upper: %v", capacity, lower, upper)
+
+		return 0.0
+	}
+
+	return p.primitive(capacity, upper) - p.primitive(capacity, lower)
+}
+
+// probabilityFormula computes the expected probability for a payment of
+// amountMsat given prior learnings for a channel of certain capacity.
+// successAmountMsat and failAmountMsat stand for the unsettled success and
+// failure amounts, respectively.
+func (p *probabilityBimodal) probabilityFormula(capacityMsat, successAmountMsat,
+	failAmountMsat, amountMsat lnwire.MilliSatoshi) float64 {
+
+	// Convert to positive-valued floats.
+	capacity := float64(capacityMsat)
+	successAmount := float64(successAmountMsat)
+	failAmount := float64(failAmountMsat)
+	amount := float64(amountMsat)
+
+	// Capacity being zero is a sentinel value to ignore the probability
+	// estimation, we'll return the full probability here.
+	if capacity == 0.0 {
+		return 1.0
+	}
+
+	// We cannot send more than the capacity.
+	if amount > capacity {
+		return 0.0
+	}
+
+	// Mission control may have some outdated values, correct them.
+	// failAmountMsat should be capacityMsat at max.
+	if failAmount > capacity {
+		failAmount = capacity
+	}
+
+	// successAmountMsat should be capacityMsat at max.
+	if successAmount > capacity {
+		successAmount = capacity
+	}
+
+	// The next statement is a safety check against an illogical condition,
+	// otherwise the renormalization integral would become zero.
+	if failAmountMsat <= successAmountMsat {
+		log.Errorf("fail amount (%v) is larger than success amount "+
+			"(%v) for capacity (%v)", failAmountMsat,
+			successAmountMsat, capacityMsat)
+
+		return 0.0
+	}
+
+	// We cannot send more than the fail amount.
+	if amount >= failAmount {
+		return 0.0
+	}
+
+	// The success proability for payment amount a is the integral over the
+	// prior distribution P(x), the probability to find liquidity between
+	// the amount a and channel capacity c (or failAmountMsat a_f):
+	// P(X >= a | X < a_f) = Integral_{a}^{a_f} P(x) dx
+	prob := p.integral(capacity, amount, failAmount)
+
+	// If we have payment information, we need to adjust the prior
+	// distribution P(x) and get the posterior distribution by renormalizing
+	// the prior distribution in such a way that the probability mass lies
+	// between a_s and a_f.
+	reNorm := p.integral(capacity, successAmount, failAmount)
+
+	// A zero division should not occur here having checked the input
+	// values. We don't prevent possible zero division here to catch these
+	// cases in fuzz testing.
+	prob /= reNorm
+
+	// Note that for payment amounts smaller than successAmountMsat, we get
+	// a value larger than unity, which is capped here to get a proper
+	// probability.
+	if prob < 0 {
```

Comment:
q: should we log if prob < 0 / > 1 and payment amount is not smaller than `successAmountMsat`?